File: 프로그래머스/lv3/최적의_행렬곱셈/main.cpp
#include <string>
#include <vector>
#include <iostream>
#include <memory.h>
#include <algorithm>

using namespace std;

int cache[200][200]={0,};

int dfs(int l, int r, vector<vector<int>> &ms)
{
	if(l==r) return 0;

	if(cache[l][r] != -1) return cache[l][r];
	cache[l][r] = 0x01<<30;
	
	for(int i = l ; i < r ; i++){
		cache[l][r] = min( cache[l][r],  dfs(l, i, ms) + dfs(i+1, r, ms) + ms[l][0] * ms[i][1] * ms[r][1] );
	}
	return cache[l][r];
}

int solution(vector<vector<int>> matrix_sizes) {
	vector<vector<int>> &ms = matrix_sizes;
	memset(cache, -1, sizeof(cache));
	int N = matrix_sizes.size();
	//접근방법 1 2d dynamic programming
	// cache[p][q] = p부터 q까지의 행렬곱 최소 연산 수
	// cache[p][q] = min(  cache[p][q], X);

    return dfs(0, N - 1, matrix_sizes);
}

File: 프로그래머스/lv3/최적의_행렬곱셈/main.java
import java.util.*;

class Solution {
    int INF = Integer.MAX_VALUE;
    int answer = 0;
    int[][] cache;
    
    int dfs(int left, int right, int [][] matrixSizes) {
        if(left == right) {
            return 0;
        }
        
        if(cache[left][right] != -1) {
            return cache[left][right];
        }
        
        cache[left][right] = 0x01<<30;
        
        for(int i = left ; i < right ; i++) {
            cache[left][right] = Math.min(
                cache[left][right],
                dfs(left, i, matrixSizes) + dfs(i+1, right, matrixSizes) + matrixSizes[left][0] * matrixSizes[i][1] * matrixSizes[right][1]
            );
        }
        
        return cache[left][right];
    }
    
    public int solution(int[][] matrixSizes) {
        this.cache = new int[matrixSizes.length][matrixSizes.length];
        
        for(int i = 0 ; i < cache.length ; i ++) {
            for(int j = 0  ; j < cache.length ; j++) {
                cache[i][j] = -1;
            }
        }
        
        answer = dfs(0, matrixSizes.length - 1, matrixSizes);
        
        return answer;
    }
}